src/preprocess.c

```diff
@@ -0,0 +1,338 @@
+/*
+ * Created by Yotam Levit on 24/07/2024.
+*/
+
+#include "../include/logger.h"
+#include "../include/preprocess.h"
+#include "../include/utils.h"
+#include <stdio.h>
+#include <string.h>
+
+
+#define MACRO_START "macr"
+#define MACRO_END "endmacr\n"
+#define EOS '\0'
+#define NEW_LINE '\n'
+#define ASM_COMMENT ';'
+#define MACRO_OUTPUT_EXTENSION "asm"
+#define SPACE " "
+
+#define REMOVE_NEW_LINE(str) *strchr(str, NEW_LINE) = EOS
+
+
+/**
+ * Checks if a given position in a string is the start of a macro definition.
+ *
+ * @param pos The position in the string to check.
+ * @return A boolean indicating if the position is the start of a macro definition.
+ */
+boolean is_macro_definition(char* pos) {
+    return strncmp(pos, MACRO_START, strlen(MACRO_START)) == 0;
+}
+
+
+/**
+ * Counts the number of macro occurrences in a file.
+ *
+ * @param file The file to count macro occurrences in.
+ * @return The number of macro occurrences.
+ */
+int count_macro_occurrences(FILE* file) {
+    char buffer[MAX_LINE_LENGTH];
+    char* pos;
+    int count = 0;
+
+    while (fgets(buffer, MAX_LINE_LENGTH, file) != NULL) {
+        pos = delete_first_spaces(buffer);
+        strtok(pos, SPACE);
+
+        if (is_macro_definition(buffer))
+            count++;
+    }
+
+    return count;
+}
+
+
+/**
+ * Initializes a hash map for storing macros.
+ *
+ * @param file The file to read macros from.
+ * @return A pointer to the initialized hash map.
+ */
+HashMapPtr init_macro_hash_map(FILE* file) {
+    return createHashMap(count_macro_occurrences(file), NULL, NULL);
+}
+
+
+/**
+ * Allocates memory for an object of a given size.
+ *
+ * @param object_size The size of the object to allocate memory for.
+ * @return A pointer to the allocated memory.
+ */
+void *handle_malloc(long object_size) {
+    void *object_ptr = malloc(object_size);
+    if (object_ptr == NULL) {
+        error_log("ERROR_CODE_1");
```

Comment:
I think there can be a more detailed error
```suggestion
        error_log("Failed to allocate memory");
```